sw/device/tests/sram_ctrl_execution_test_ret.c

```diff
@@ -0,0 +1,178 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/csr.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/stdasm.h"
+#include "sw/device/lib/dif/dif_sram_ctrl.h"
+#include "sw/device/lib/handler.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/sram_ctrl_testutils.h"
+#include "sw/device/lib/testing/test_framework/test_main.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+const test_config_t kTestConfig;
+
+static dif_sram_ctrl_t sram_ctrl;
+
+/**
+ * This flag is used to verify that the instruction access fault has occurred,
+ * and has been successfully serviced. Declared as volatile, because it is
+ * referenced in the fault handler, as well as the main test flow.
+ */
+static volatile bool instruction_access_fault;
+
+static const uint32_t kRetRamStartAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR;
+static const uint32_t kRetRamEndAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR +
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_SIZE_BYTES - 1;
+
+/**
+ * Buffer filled with `0` (valid `unimp`) instructions.
+ *
+ * When Retention SRAM is disabled, executing this buffer will result in an
+ * Instruction Access Fault, which constitutes a successful test. Otherwise,
+ * it will result in an Illegal Instruction Fault, which constitutes a failure.
+ */
+static const uint32_t execution_test_instructions[2];
+
+/**
+ * Determines and sets the return address from an exception handler.
+ *
+ * The exception cause instruction, could be either standard or compressed,
+ * which in turn means that the return address (next instruction after fault)
+ * will be either at the 2byte (16bit) or 4byte (32bit) offset.
+ *
+ * @param indirection `NULL` to calculate return address relative to the
+ *                    exception cause instruction (default), or an alternative
+ *                    return address.
+ */
+static void fault_handler_set_return_addr(uintptr_t *indirection) {
```

Comment:
This is never called with a NULL value. Could delete the `indirection == NULL` path and make `indirection` a value instead of a pointer.